linalg/sparsemat.cpp

```diff
@@ -2461,52 +2461,67 @@ void SparseMatrix::DiagScale(const Vector &b, Vector &x, double sc) const
    });
 }
 
-void SparseMatrix::Jacobi2(const Vector &b, const Vector &x0, Vector &x1,
-                           double sc) const
+template <bool useFabs>
+static void JacobiDispatch(const Vector &b, const Vector &x0, Vector &x1,
+                           const Memory<int> &I, const Memory<int> &J,
+                           const Memory<double> &A, const int height,
+                           const double sc)
 {
-   MFEM_VERIFY(Finalized(), "Matrix must be finalized.");
+   const bool useDevice = b.UseDevice() || x0.UseDevice() || x1.UseDevice();
 
-   for (int i = 0; i < height; i++)
+   const auto bp  = b.Read(useDevice);
+   const auto x0p = x0.Read(useDevice);
+   auto       x1p = x1.Write(useDevice);
+
+   const auto Ip = Read(I, height+1, useDevice);
+   const auto Jp = Read(J, J.Capacity(), useDevice);
+   const auto Ap = Read(A, J.Capacity(), useDevice);
+
+   MFEM_FORALL_SWITCH(useDevice, i, height,
```

Comment:
All other `SparseMatrix` methods that support devices use `MFEM_FORALL` and not `MFEM_FORALL_SWITCH`, i.e. they follow the global configuration about where to run. It may be better to follow the same logic here.

Reply:
> all other SparseMatrix methods that support devices use MFEM_FORALL and not MFEM_FORALL_SWITCH

It isn't consistent. `SparseMatrix::DiagScale()` (immediately above `JacobiDispatch()`) for example uses `MFEM_FORALL_SWITCH`, so I went with the more flexible option. Either way, this routine should still respect the global configuration as both the input `Vector`'s should be on host if the user does not enable device, in which case this routine will also run on host.

Reply:
Okay, I had not seen the `MFEM_FORALL_SWITCH` in `DiagScale` -- I'm not sure why it was used there.

My concern is that since general `Vector`s are not automatically marked to use device, using `MFEM_FORALL_SWITCH` in these `SparseMatrix` methods, one may end up running them on host even if the global setting is to run on GPU.

The reason for using `MFEM_FORALL_SWITCH` in `Vector` operations was to make sure small-sized `Vector`s are not pushed unnecessarily on GPU -- also that is why `Vector`s are NOT marked to use the device by default. With `SparseMatrix`, we don't have small-size `SparseMatrix`es that we want to force to stay on CPU, so the intent for `SparseMatrix` is to always run on GPU if that is the configured device.

In any case, this is something that can be corrected in the future, if we notice it to be an issue -- no need to hold the PR over this.